src/cartesian.rs

```diff
@@ -0,0 +1,227 @@
+use std::iter::Sum;
+use std::mem;
+use std::ops::Range;
+
+use num_traits::AsPrimitive;
+use num_traits::Num;
+
+#[derive(Copy, Clone, Debug)]
+pub struct Grid<const D: usize> {
+    size: [usize; D],
+}
+
+impl<const D: usize> Grid<D> {
+    fn into_subgrid(self) -> SubGrid<D> {
+        SubGrid {
+            size: self.size,
+            offset: [0; D],
+        }
+    }
+
+    fn position_of(&self, mut i: usize) -> [usize; D] {
+        let mut pos = [0; D];
+        for (s, p) in self.size.into_iter().zip(&mut pos) {
+            *p = i % s;
+            i /= s;
+        }
+        pos
+    }
+}
```

Comment:
Grids are not necessarily squares (n x n) or cubes (n x n x n) but it seems to me that this hypothesis is leveraged in this function, no?

Reply:
Is there something wrong I'm not seeing?

For a 2D grid of size `[W,H]`, the position of the `i`th element in a row-major array is `[i%W,I/W]`:

```
       <---- W ---->
^  0   1   2  ...  W-2  W-1
|  W  W+1 W+2 ... 2W-2 2W-1
H  *   *   *  ...   *   *
|  ........... i .........
v  *   *   *  ...   *  WH-1
```

For a 3D grid of size `[W,H,D]`, it's `[i%W,(i/W)%H,i/W/H]`.

The function computes the last component as `(i/W)%H` for 2D or `(i/W/H)%D` for 3D, but it doesn't change the result for `i`  smaller than `W*H` or `W*H*D`.

Reply:
You are right i did misread the function :+1: